### DISCUSSION TITLE
Tab completion of standard directories and "z -" not working

### DISCUSSION THREAD

**Discussion by alewis001** [ajeetdsouza/zoxide#835] [Q&A] [answered]
Hi.

I've searched the issues and discussions and although I see items around Tab completion, I'm not sure they're the same issue/behaviour that I'm experiencing. Apologies if I end up asking a duplicate.

I have the impression from YouTube videos I've watched and things I've read that I should be able to do `z <start of some path>` and hit tab and see the same tab completion that I'd normally get when using `cd`. For instance, if I do `z ~/<tab>` nothing happens currently. However, if I do `z <some character>`, I do get tab completion for zoxide paths.

Do I have the wrong expectations or does it look like something is wrong in my setup?

As an aside but maybe related, doing `z -` doesn't switch to the last dir, instead it gives me a message about Improper option(s) given and help on the command line args.

I'm using zsh (with oh-my-zsh), in tmux, in iterm2 on macos. I've added the `eval "$(zoxide init zsh)"` line as the very last line of my `.zshrc`; I also did the completion cache clear as mentioned in the instructions.

Versions:
  - macos: Sonoma 14.4.1 (23E224)
  - zsh: zsh 5.9 (x86_64-apple-darwin23.0)
  - zoxde: 0.9.4

Any help would be greatly appreciated as I would like to regularly use this tool as it looks to be very useful :)

**Comment by alewis001** [comment, marked as answer]
As an update to this... I changed the eval to this `eval "$(zoxide init --cmd cd zsh)"` and cd tab completion appears to be working as normal but I think normal cd'ing is now recording results into the `z` history. So I think the behaviour I was expecting above was a misunderstanding on my part. I was expecting that using `z` directly would work exactly the same as `cd` from a tab completion point-of-view but instead you have to use the `--cmd cd` and have `z` replace `cd` to then use `cd` and `z` together. 

I think what may have mislead me was watching videos of `z` being used and I thought I could see the `z` tab completion behaviour that I expected (I.e. the same as cd) but, without actually _seeing_ the keys the user was pressing, they may not having been pressing tab and just pasting in the paths they wanted 🤷

**Reply by JamesJosephFinn** [reply]
I may be mistaken, but I'm pretty sure there was an older utility named `z` which did have this tab completion functionality, but its popularity was overtaken by zoxide. It absolutely would be nice to have that. Since you're a Mac user, you'll know about the `SFT+CMD+G` functionality in Finder. It's perfect: tab completion, fuzzy/typo matching ... but it's not in the CLI, obviously. I'm as surprised as you that fuzzy tab expansion is not part of zoxide, given its popularity; and given that it replaces another utility with the same command (namely z) that did have fuzzy expansion.

**Reply by ajeetdsouza** [reply]
Not sure what you mean by fuzzy expansion, but zoxide's `z` command does have regular `cd` completions.

**Reply by awe123343** [reply]
@ajeetdsouza it's not working for me as well. `z` then `tab` only shows candidates from history I have z-ed into. `z ./` then `tab` actually showed me nothing, I have double checked my `~/.zshrc` and it's in the right location

**Reply by JamesJosephFinn** [reply]
@ajeetdsouza firstly, thank you for sharing this great tool. I use it every day, and find it very helpful. Thank you also for the prompt reply.

To be clear, typing `z app`, for example, followed by `SPACE` and then `TAB` opens the `fzf` TUI; which is the same as running `zi app` and hitting `ENTER`. However, typing `z app` and immediately pressing `TAB` does nothing; whereas the behavior of `cd` in this manner is to show case-sensitive completions/suggestions inline.

The old `z` from back in the day retained this `cd` inline completion/suggestion behavior, with the added obvious benefit of being typo-tolerant and case-insensitive.

I hope that all makes sense!

**Reply by awe123343** [reply]
Actually I figured it out, it's due to the `z` plugin installed by `oh-my-zsh`, once I commented it out, the completion works just like a charm.
@JamesJosephFinn not sure if it's same cause for you, worth checking tho.

**Reply by JamesJosephFinn** [reply]
> Actually I figured it out, it's due to the `z` plugin installed by `oh-my-zsh`, once I commented it out, the completion works just like a charm. @JamesJosephFinn not sure if it's same cause for you, worth checking tho.

I'm not using `oh-my-zsh`. Glad to see you've figured it out, though!

**Reply by ivancuric** [reply]
I'd like to follow up on this. 

I am only using
```zsh
# junegunn/fzf-bin
zinit ice from"gh-r" as"program"
zinit light junegunn/fzf

eval "$(mise activate zsh)"
# zinit light ajeetdsouza/zoxide // DOES NOT WORK
eval "$(zoxide init zsh)"
```

Using autocomplete with space+tab, I also only get matches from the history, which is not really helpful since I am constantly making new git worktrees.

Is there any way to have a fuzzy search for all nested directories, not just ones in history?
